### PR TITLE
Consolidate SHAMap::addRootNode overrides.

### PR DESCRIPTION
**Pull request by ximinez**
* Remove SHAMapAbstractNode::dump
* Clean up some formatting.

Reviewers: @HowardHinnant, @miguelportilla

### CODE REVIEW

File: src/ripple/shamap/SHAMap.h [hunk 1]
```diff
@@ -182,10 +182,9 @@ class SHAMap
     
     bool getRootNode (Serializer & s, SHANodeFormat format) const;
     std::vector<uint256> getNeededHashes (int max, SHAMapSyncFilter * filter);
-    SHAMapAddNode addRootNode (uint256 const& hash, Blob const& rootNode, SHANodeFormat format,
-                               SHAMapSyncFilter * filter);
-    SHAMapAddNode addRootNode (Blob const& rootNode, SHANodeFormat format,
-                               SHAMapSyncFilter * filter);
+    SHAMapAddNode addRootNode (boost::optional<uint256> const& hash,
```
Comment (vinniefalco) [inline src/ripple/shamap/SHAMap.h]: Are you sure we should be passing `optional`?

Comment (ximinez) [inline src/ripple/shamap/SHAMap.h]: `optional` seems the simplest way to be able to indicate that the caller doesn't have a `hash`. If there's a better way, I'm open to it.

Comment (vinniefalco) [inline src/ripple/shamap/SHAMap.h]: Sorry, brain moment - I meant are you sure we should pass `const&` and not by value?

Comment (ximinez) [inline src/ripple/shamap/SHAMap.h]: I don't think it makes much difference here.None of the callers pass in a `boost::optional`, so it's going to have to make one either way. Change it?

Comment (HowardHinnant) [inline src/ripple/shamap/SHAMap.h]: `const&` will have better performance.

**Comment by miguelportilla** [comment]
LGTM :+1:

**Comment by HowardHinnant** [comment]
:+1:

File: src/ripple/shamap/SHAMap.h (hunk 1 above)
Comment (vinniefalco) [inline src/ripple/shamap/SHAMap.h]: I studied the body of `addRootNote` and I don't understand why/how the `hash` is optional. What does it mean when you leave it out? Where is it used?

Comment (HowardHinnant) [inline src/ripple/shamap/SHAMap.h]: There are currently two `addRootNode` overloads with identical signatures except one is missing a `hash` argument.  The definitions of these two functions are nearly identical.  The one without the `hash` is called only from a test in SHAMapSync.test.cpp.
This P/R simply consolidates the two overloads into one, eliminating some repeated code.  The test now fills in the `hash` with `boost::none`.

Comment (HowardHinnant) [inline src/ripple/shamap/SHAMap.h]: Additionally the `hash`, when passed, is redundant information that the non-test code double-checks and reacts to with asserts (in debug) or returning error codes (in release).  The real code calls this function with the hash of a ledger.  The shamap unittest test lacks a ledger to build from.

**Comment by vinniefalco** [comment]
Production interfaces should not have unit-test specific parameters (for example `TxnSignApiFacade` appearing in the parameter list to `transactionSign`. This consolidation is not a cleanup at all since it puts unit test code into a production function.

:-1:

**Comment by ximinez** [comment]
That's fine by me, as long as we're aware that the unit test is not testing the production function. Perhaps the best solution is to rewrite the test.

**Comment by vinniefalco** [comment]
The problem is we have a function that takes a `boost::optional` that only has meaning for a test. Someone who sees the function signature might wonder, what are the semantics of the call when the parameter is `boost::none`? We should eliminate the anxiety that might confuse readers. A similar situation exists where `uint256` values can mean special things when they are zeroed.

**Comment by ximinez** [comment]
Agree.

**Pull request closed without merging**